go/libraries/doltcore/sqle/setalgebra/intersection.go

```diff
@@ -0,0 +1,240 @@
+// Copyright 2020 Liquidata, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package setalgebra
+
+import (
+	"github.com/liquidata-inc/dolt/go/store/hash"
+	"github.com/liquidata-inc/dolt/go/store/types"
+)
+
+// finiteSetIntersection returns the set of points that are in both fs1 and fs2
+func finiteSetIntersection(fs1, fs2 FiniteSet) (Set, error) {
+	hashToVal := make(map[hash.Hash]types.Value, len(fs1.HashToVal)+len(fs2.HashToVal))
+	for h, v := range fs1.HashToVal {
+		if _, ok := fs2.HashToVal[h]; ok {
+			hashToVal[h] = v
+		}
+	}
+
+	if len(hashToVal) == 0 {
+		return EmptySet{}, nil
+	} else {
+		return FiniteSet{HashToVal: hashToVal}, nil
+	}
+}
+
+// finiteSetInterval will return the set of points that are in the interval, or an EmptySet instance
+func finiteSetIntervalIntersection(fs FiniteSet, in Interval) (Set, error) {
+	hashToVal := make(map[hash.Hash]types.Value, len(fs.HashToVal))
+	for h, v := range fs.HashToVal {
+		inRange, err := ValueInInterval(in, v)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if inRange {
+			hashToVal[h] = v
+		}
+	}
+
+	if len(hashToVal) == 0 {
+		return EmptySet{}, nil
+	} else {
+		return FiniteSet{HashToVal: hashToVal}, nil
+	}
+}
+
+// finiteSetInterval will return the set of points that are in the composite set, or an EmptySet instance
+func finiteSetCompositeSetIntersection(fs FiniteSet, composite CompositeSet) (Set, error) {
```

Comment:
I actually tried this.  Because you still need to loop over all the comp.Intervals unioning all the results.  with the error handling on each of the unions and intersections the code actually comes out looking worse imo.